### ISSUE TITLE
MIGRAPHX_TRACE_MLIR should dump out the constructed MLIR before benchmarking / quick tuning

### ISSUE THREAD

**Issue by krzysz00** [ROCmSoftwarePlatform/AMDMIGraphX#2332]
It's hard to debug rocMLIR issues right now because the trace variable is only effective during the final run, but not during the benchmarking/quick tuning stage, which is when we might want a look at the generated MLIR so that we can, for example, work out why shufflenet is crashing.

**Comment by manupak** [comment]
Updated the ticket @jerryyin @krzysz00

**Comment by pfultz2** [comment]
> Fix P2 : We need a another dump /w MIGRAPHX_TRACE_MLIR=1 before it is benchmarked along with the solution.

I dont think we can dump MLIR after its been compiled. The benchmarking only has the binaries.

What we can do is print out the solutions key in the mlir trace dumps, and then during benchmarking with something like`MIGRAPHX_TRACE_BENCHMARK=1`, we can show the solution key it is trying to run. Then you can cross reference it to get the MLIR program.

**Comment by manupak** [comment]
> I dont think we can dump MLIR after its been compiled. The benchmarking only has the binaries.

I dont understand why this is not possible. MIGraphX just need to carry a span reference back to the migraphx module until after the benchmark -- probably in the compiled_result

> What we can do is print out the solutions key in the mlir trace dumps, and then during benchmarking with something like MIGRAPHX_TRACE_BENCHMARK=1, we can show the solution key it is trying to run. Then you can cross reference it to get the MLIR program.

You are saying something like : 

```
Benchmarking gpu::mlir_op: 36 configs
Problem : gfx90a:sramecc+:xnack-	110	-t f16 -out_datatype f16 -transA false -transB false -g 1 -m 384 -n 2304 -k 768
Solution : 128,256,4,64,128,8,1,1
Solution : 128,256,2,64,128,8,1,1
```
right ? 

well this is not good enough because its a one to many map to MLIR programs starting the from that problem key ( + solution).

Anyhow, -- correct me @jerryyin if I am wrong --I think we are trying to get to a point for MIGraphX team to raise bugs (tickets) saying which MLIR program hangs/fails/slow rather than rocMLIR trying to debug the model failure starting the MIGraphX.

So,
 
> Then you can cross reference it to get the MLIR program

If you think above is doable from MIGraphX team, I personally dont have issues; just that I dont see how one could do that when multiple (different) MLIR programs have the same problem key.

**Comment by umangyadav** [comment]
When you set `MIGRAPHX_TRACE_BENCHMARKING=3` it would print MLIR module before benchmarking 

https://github.com/ROCm/AMDMIGraphX/blob/3e65032266c013e840fd08a2f41623a822e1c538/src/targets/gpu/compile_ops.cpp#L207

https://github.com/ROCm/AMDMIGraphX/blob/3e65032266c013e840fd08a2f41623a822e1c538/src/targets/gpu/compile_ops.cpp#L86

https://github.com/ROCm/AMDMIGraphX/blob/3e65032266c013e840fd08a2f41623a822e1c538/src/targets/gpu/jit/mlir.cpp#L125

https://github.com/ROCm/AMDMIGraphX/blob/3e65032266c013e840fd08a2f41623a822e1c538/src/targets/gpu/jit/mlir.cpp#L207

**Comment by umangyadav** [comment]
For the split-k fusion logic part, it is not printing right now but that can be added easily 
https://github.com/ROCm/AMDMIGraphX/blob/3e65032266c013e840fd08a2f41623a822e1c538/src/targets/gpu/jit/mlir.cpp#L137

**Comment by umangyadav** [comment]
> Fix P1 : We need a dump /w MIGRAPHX_TRACE_MLIR=1 of migraphx module and MLIR module before it gets compiled for the first time.
Fix P2 : We need a another dump /w MIGRAPHX_TRACE_MLIR=1 before it is benchmarked along with the solution. 

we already have Fix P1 i think. 
We have partial fix P2.
